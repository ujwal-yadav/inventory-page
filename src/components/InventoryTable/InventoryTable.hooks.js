import axios from 'axios';
import { useAuth } from '../../context/AuthProvider';
import { useEffect, useState } from 'react';

export const useInventoryTable = () => {
  const [inventoryData, setInventoryData] = useState(null);
  const [isLoading, setIsLoading] = useState(false);
  const [isEditModalOpen, setIsEditModalOpen] = useState(false);
  const [isDeleteModalOpen, setIsDeleteModalOpen] = useState(false);
  const [activeProductId, setActiveProductId] = useState(null);
  const [errors, setErrors] = useState({
    name: false,
    price: false,
    brand: false,
    category: false,
    stock: false,
  });

  const [product, setProduct] = useState({
    name: '',
    price: '',
    brand: '',
    category: '',
    stock: '',
  });

  const validateField = (name, value) => {
    setErrors((prevErrors) => ({
      ...prevErrors,
      [name]: !value,
    }));

    return !value;
  };

  const handleChange = (e) => {
    const { name, value } = e.target;
    setProduct((prevProduct) => ({
      ...prevProduct,
      [name]: value,
    }));
    validateField(name, value);
  };

  const { token } = useAuth();

  const headers = {
    token,
  };

  const getProducts = async () => {
    setIsLoading(true);

    try {
      const response = await axios.get('/products', { headers });

      if (response.data) {
        setInventoryData(response.data.data);
      }
    } catch (err) {
      alert('Error while fetching inventory details!');
    } finally {
      setIsLoading(false);
    }
  };

  useEffect(() => {
    getProducts();
  }, []);

  const toggleEditProductModal = (id) => {
    setActiveProductId(id);
    setIsEditModalOpen(!isEditModalOpen);
    if (!isEditModalOpen && inventoryData) {
      const productData = inventoryData.find((product) => product.id === id);
      setProduct({
        name: productData.name,
        price: productData.price,
        brand: productData.brand.code,
        category: productData.product_category.id,
        stock: productData.minimum_stock,
      });
    }
  };

  const toggleDeleteProductModal = (id) => {
    setActiveProductId(id);
    setIsDeleteModalOpen(!isDeleteModalOpen);
  };

  const deleteProduct = async () => {
    setIsDeleteModalOpen(false);
    try {
      const response = await axios.delete(`/products/${activeProductId}`, {
        headers,
      });

      if (response.data === '') {
        getProducts();
      }
    } catch (err) {
      alert('Could not delete product!');
    }
  };

  const editProduct = async () => {
    const { name, price, brand, category, stock } = product;
    const nameError = validateField('name', name);
    const priceError = validateField('price', price);
    const brandError = validateField('brand', brand);
    const categoryError = validateField('category', category);
    const stockError = validateField('stock', stock);

    if (nameError || priceError || brandError || categoryError || stockError) {
      return;
    }

    try {
      const response = await axios.put(
        `/products/${activeProductId}`,
        {
          id: activeProductId,
          name,
          price: Number(price),
          brand: String(brand),
          product_category: String(category),
          minimum_stock: String(stock),
        },
        { headers }
      );

      if (response.data) {
        setIsEditModalOpen(false);
        getProducts();
      }
    } catch (err) {
      alert('Could not update product!');
    }
  };

  return {
    isLoading,
    inventoryData,
    isEditModalOpen,
    isDeleteModalOpen,
    toggleEditProductModal,
    toggleDeleteProductModal,
    deleteProduct,
    editProduct,
    activeProductId,
    errors,
    validateField,
    handleChange,
    ...product,
  };
};
