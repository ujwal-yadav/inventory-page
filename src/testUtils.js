import { createTheme, ThemeProvider } from '@mui/material';
import { render } from '@testing-library/react';

export const renderWithTheme = (ui, { themeOptions = {} } = {}) => {
  const theme = createTheme(themeOptions);
  return render(<ThemeProvider theme={theme}>{ui}</ThemeProvider>);
};
