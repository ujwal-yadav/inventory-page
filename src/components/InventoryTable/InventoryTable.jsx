import * as React from 'react';
import { styled } from '@mui/material/styles';
import Table from '@mui/material/Table';
import TableBody from '@mui/material/TableBody';
import TableCell, { tableCellClasses } from '@mui/material/TableCell';
import TableContainer from '@mui/material/TableContainer';
import TableHead from '@mui/material/TableHead';
import TableRow from '@mui/material/TableRow';
import Paper from '@mui/material/Paper';
import {
  Box,
  Button,
  Modal,
  TextField,
  Typography,
  useMediaQuery,
} from '@mui/material';
import { useInventoryTable } from './InventoryTable.hooks';
import ModeEditIcon from '@mui/icons-material/ModeEdit';
import DeleteIcon from '@mui/icons-material/Delete';
import { IconButton } from '@mui/material';
import ClearIcon from '@mui/icons-material/Clear';
import { useTheme } from '@emotion/react';

export default function InventoryTable() {
  const theme = useTheme();
  const isLargeView = useMediaQuery(theme.breakpoints.up('md'));

  const {
    isLoading,
    inventoryData,
    isEditModalOpen,
    isDeleteModalOpen,
    toggleEditProductModal,
    toggleDeleteProductModal,
    deleteProduct,
    editProduct,
    errors,
    handleChange,
    name,
    price,
    brand,
    category,
    stock,
  } = useInventoryTable();

  if (isLoading)
    return (
      <Typography variant="h6" align="center">
        Loading...
      </Typography>
    );

  if (!inventoryData?.length)
    return (
      <Typography variant="h6" align="center">
        No Products available
      </Typography>
    );

  return (
    <>
      <Modal open={isDeleteModalOpen} onClose={toggleDeleteProductModal}>
        <Box
          sx={{
            position: 'absolute',
            top: '50%',
            left: '50%',
            transform: 'translate(-50%, -50%)',
            width: isLargeView ? 400 : '85%',
            bgcolor: 'background.paper',
            boxShadow: 24,
            p: 3,
            borderRadius: 2,
          }}
        >
          <Typography variant="h5" align="center" fontWeight={'600'}>
            Delete Product
          </Typography>
          <IconButton
            sx={{ position: 'absolute', top: 0, right: 0, margin: 2 }}
            onClick={toggleDeleteProductModal}
            size="small"
          >
            <ClearIcon />
          </IconButton>

          <Typography
            variant="h6"
            align="center"
            fontWeight={'400'}
            sx={{ paddingTop: 2 }}
            color={'error'}
          >
            Are you sure you want to delete the product ?
          </Typography>
          <Button
            sx={{
              boxShadow: 0,
              borderRadius: 2,
              fontWeight: 600,
              marginTop: 2,
            }}
            variant="contained"
            color="error"
            fullWidth
            onClick={deleteProduct}
          >
            Delete
          </Button>
        </Box>
      </Modal>

      <Modal open={isEditModalOpen} onClose={toggleEditProductModal}>
        <Box
          sx={{
            position: 'absolute',
            top: '50%',
            left: '50%',
            transform: 'translate(-50%, -50%)',
            width: isLargeView ? 400 : '85%',
            bgcolor: 'background.paper',
            boxShadow: 24,
            p: 3,
            borderRadius: 2,
          }}
        >
          <Typography
            variant="h5"
            align="center"
            fontWeight={'600'}
            sx={{ paddingBottom: 1 }}
          >
            Edit Product
          </Typography>
          <IconButton
            sx={{ position: 'absolute', top: 0, right: 0, margin: 2 }}
            onClick={toggleEditProductModal}
            size="small"
          >
            <ClearIcon />
          </IconButton>

          <TextField
            margin="normal"
            required
            fullWidth
            id="name"
            label="Product Name"
            name="name"
            autoFocus
            error={errors.name}
            helperText={errors.name ? 'Product name is required' : ''}
            value={name}
            onChange={handleChange}
            InputProps={{
              style: {
                borderRadius: '8px',
              },
            }}
          />
          <TextField
            margin="normal"
            required
            fullWidth
            name="price"
            label="Product Price"
            id="price"
            type="number"
            error={errors.price}
            helperText={errors.price ? 'Product price is required' : ''}
            value={price}
            onChange={handleChange}
            InputProps={{
              style: {
                borderRadius: '8px',
              },
            }}
          />
          <TextField
            margin="normal"
            required
            fullWidth
            name="brand"
            label="Brand"
            id="brand"
            error={errors.brand}
            helperText={errors.brand ? 'Brand is required' : ''}
            value={brand}
            onChange={handleChange}
            InputProps={{
              style: {
                borderRadius: '8px',
              },
            }}
          />
          <TextField
            margin="normal"
            required
            fullWidth
            name="category"
            label="category"
            id="category"
            error={errors.category}
            helperText={errors.category ? 'Product Category is required' : ''}
            value={category}
            onChange={handleChange}
            InputProps={{
              style: {
                borderRadius: '8px',
              },
            }}
          />
          <TextField
            margin="normal"
            required
            fullWidth
            name="stock"
            label="Stock"
            id="stock"
            error={errors.stock}
            helperText={errors.stock ? 'Stock is required' : ''}
            value={stock}
            onChange={handleChange}
            InputProps={{
              style: {
                borderRadius: '8px',
              },
            }}
          />

          <Button
            sx={{
              boxShadow: 0,
              borderRadius: 2,
              fontWeight: 600,
              marginTop: 2,
            }}
            variant="contained"
            color="info"
            fullWidth
            onClick={editProduct}
          >
            Edit
          </Button>
        </Box>
      </Modal>

      <Typography
        variant="h5"
        component="h1"
        sx={{ paddingBottom: 2, fontWeight: 600 }}
      >
        Inventory
      </Typography>
      <TableContainer component={Paper} sx={{ borderRadius: 2 }}>
        <Table sx={{ minWidth: 700 }} aria-label="customized table">
          <TableHead>
            <TableRow>
              <StyledTableCell>Brand Code</StyledTableCell>
              <StyledTableCell>Brand Name</StyledTableCell>
              <StyledTableCell>Product Code</StyledTableCell>
              <StyledTableCell>Product Name</StyledTableCell>
              <StyledTableCell>Product Category</StyledTableCell>
              <StyledTableCell>Stock</StyledTableCell>
              <StyledTableCell>Price</StyledTableCell>
              <StyledTableCell
                sx={{
                  position: 'sticky',
                  right: 0,
                  top: 0,
                  background: 'white',
                  justifyContent: 'center',
                }}
              ></StyledTableCell>
            </TableRow>
          </TableHead>
          <TableBody>
            {inventoryData?.map((row) => (
              <StyledTableRow key={row.id}>
                <StyledTableCell>{row.brand.code}</StyledTableCell>
                <StyledTableCell>{row.brand.name}</StyledTableCell>
                <StyledTableCell>{row.code}</StyledTableCell>
                <StyledTableCell>{row.name}</StyledTableCell>
                <StyledTableCell>{row.product_category.name}</StyledTableCell>
                <StyledTableCell>{row.minimum_stock}</StyledTableCell>
                <StyledTableCell>â‚¹{row.price}</StyledTableCell>
                <StyledTableCell
                  sx={{
                    position: 'sticky',
                    right: 0,
                    background: 'white',
                    justifyContent: 'center',
                    display: 'flex',
                    gap: 2,
                    padding: '20px 25px',
                  }}
                >
                  <IconButton onClick={() => toggleEditProductModal(row.id)}>
                    <ModeEditIcon color="info" />
                  </IconButton>
                  <IconButton onClick={() => toggleDeleteProductModal(row.id)}>
                    <DeleteIcon color="error" />
                  </IconButton>
                </StyledTableCell>
              </StyledTableRow>
            ))}
          </TableBody>
        </Table>
      </TableContainer>
    </>
  );
}

const StyledTableRow = styled(TableRow)(({ theme }) => ({
  '&:nth-of-type(odd)': {
    backgroundColor: theme.palette.action.hover,
  },
  '&:last-child td, &:last-child th': {
    border: 0,
  },
}));

const StyledTableCell = styled(TableCell)(({ theme }) => ({
  [`&.${tableCellClasses.head}`]: {
    backgroundColor: theme.palette.primary.main,
    color: theme.palette.common.white,
    fontWeight: '550',
    [theme.breakpoints.up('md')]: {
      fontSize: 20,
      padding: 25,
    },
  },
  [`&.${tableCellClasses.body}`]: {
    [theme.breakpoints.up('md')]: {
      fontSize: 20,
      padding: '20px 25px',
    },
  },
}));
