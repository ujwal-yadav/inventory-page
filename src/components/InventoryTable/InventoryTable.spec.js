import { render, screen } from '@testing-library/react';
import { useInventoryTable } from './InventoryTable.hooks';
import InventoryTable from './InventoryTable';
import { renderWithTheme } from './../../testUtils';

jest.mock('./InventoryTable.hooks', () => ({
  useInventoryTable: jest.fn(),
}));

describe('InventoryTable', () => {
  it('should display "Loading..." when isLoading is true', () => {
    useInventoryTable.mockReturnValue({ isLoading: true, inventoryData: null });

    render(<InventoryTable />);

    expect(screen.getByText('Loading...')).toBeInTheDocument();
  });

  it('should display "No Products available" when inventoryData is empty', () => {
    useInventoryTable.mockReturnValue({ isLoading: false, inventoryData: [] });

    render(<InventoryTable />);

    expect(screen.getByText('No Products available')).toBeInTheDocument();
  });

  it('should render the product details correctly', () => {
    const mockData = [
      {
        id: 1,
        code: 'PROD-01',
        name: 'Product Satu',
        price: 1000,
        minimum_stock: 25,
        company: {
          id: 1,
          code: 'DM',
          name: 'Dummy',
          address: '',
        },
        brand: {
          id: 1,
          company: {
            id: 0,
            code: '',
            name: '',
            address: '',
          },
          code: 'BRAND-01',
          name: 'Brand Test',
        },
        product_category: {
          id: 1,
          company: {
            id: 0,
            code: '',
            name: '',
            address: '',
          },
          name: 'Furniture',
          category: {
            id: 0,
            name: '',
          },
        },
      },
    ];

    useInventoryTable.mockReturnValue({
      isLoading: false,
      inventoryData: mockData,
    });

    const { asFragment } = renderWithTheme(<InventoryTable />);

    expect(asFragment()).toMatchSnapshot();
  });
});
