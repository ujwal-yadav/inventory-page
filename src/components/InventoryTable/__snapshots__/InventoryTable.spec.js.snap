// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`InventoryTable should render the product details correctly 1`] = `
<DocumentFragment>
  <h1
    class="MuiTypography-root MuiTypography-h5 css-1wiiowb-MuiTypography-root"
  >
    Inventory
  </h1>
  <div
    class="MuiPaper-root MuiPaper-elevation MuiPaper-rounded MuiPaper-elevation1 MuiTableContainer-root css-1vdaxjy-MuiPaper-root-MuiTableContainer-root"
  >
    <table
      aria-label="customized table"
      class="MuiTable-root css-1pclhgl-MuiTable-root"
    >
      <thead
        class="MuiTableHead-root css-15wwp11-MuiTableHead-root"
      >
        <tr
          class="MuiTableRow-root MuiTableRow-head css-1q1u3t4-MuiTableRow-root"
        >
          <th
            class="MuiTableCell-root MuiTableCell-head MuiTableCell-sizeMedium css-h4536p-MuiTableCell-root"
            scope="col"
          >
            Brand Code
          </th>
          <th
            class="MuiTableCell-root MuiTableCell-head MuiTableCell-sizeMedium css-h4536p-MuiTableCell-root"
            scope="col"
          >
            Brand Name
          </th>
          <th
            class="MuiTableCell-root MuiTableCell-head MuiTableCell-sizeMedium css-h4536p-MuiTableCell-root"
            scope="col"
          >
            Product Code
          </th>
          <th
            class="MuiTableCell-root MuiTableCell-head MuiTableCell-sizeMedium css-h4536p-MuiTableCell-root"
            scope="col"
          >
            Product Name
          </th>
          <th
            class="MuiTableCell-root MuiTableCell-head MuiTableCell-sizeMedium css-h4536p-MuiTableCell-root"
            scope="col"
          >
            Product Category
          </th>
          <th
            class="MuiTableCell-root MuiTableCell-head MuiTableCell-sizeMedium css-h4536p-MuiTableCell-root"
            scope="col"
          >
            Stock
          </th>
          <th
            class="MuiTableCell-root MuiTableCell-head MuiTableCell-sizeMedium css-h4536p-MuiTableCell-root"
            scope="col"
          >
            Price
          </th>
        </tr>
      </thead>
      <tbody
        class="MuiTableBody-root css-apqrd9-MuiTableBody-root"
      >
        <tr
          class="MuiTableRow-root css-1enotpu-MuiTableRow-root"
        >
          <td
            class="MuiTableCell-root MuiTableCell-body MuiTableCell-sizeMedium css-l9k0sc-MuiTableCell-root"
          >
            BRAND-01
          </td>
          <td
            class="MuiTableCell-root MuiTableCell-body MuiTableCell-sizeMedium css-l9k0sc-MuiTableCell-root"
          >
            Brand Test
          </td>
          <td
            class="MuiTableCell-root MuiTableCell-body MuiTableCell-sizeMedium css-l9k0sc-MuiTableCell-root"
          >
            PROD-01
          </td>
          <td
            class="MuiTableCell-root MuiTableCell-body MuiTableCell-sizeMedium css-l9k0sc-MuiTableCell-root"
          >
            Product Satu
          </td>
          <td
            class="MuiTableCell-root MuiTableCell-body MuiTableCell-sizeMedium css-l9k0sc-MuiTableCell-root"
          >
            Furniture
          </td>
          <td
            class="MuiTableCell-root MuiTableCell-body MuiTableCell-sizeMedium css-l9k0sc-MuiTableCell-root"
          >
            25
          </td>
          <td
            class="MuiTableCell-root MuiTableCell-body MuiTableCell-sizeMedium css-l9k0sc-MuiTableCell-root"
          >
            â‚¹1000
          </td>
        </tr>
      </tbody>
    </table>
  </div>
</DocumentFragment>
`;
